colinfo: add version gate for storing pg_lsn types

otan opened the pull request:
We don't want mixed version clusters storing pg_lsn types, in case they need to rollback / older versions do not understand the type.

Informs #105130

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105391)
<!-- Reviewable:end -->

rafiss commented on pkg/clusterversion/cockroach_versions.go:
```diff
@@ -542,6 +542,9 @@ const (
 	V23_2_UseACRaftEntryEntryEncodings
 
+	// V23_2_PGLogicalReplication gates the use of PGLogicalReplication features.
+	V23_2_PGLogicalReplication
```
nit: no need to make a new cluster version. you can just use `clusterversion.V23_2`

otan replied:
there is no `V23_2` though, only `V23_2Start`.

otan replied:
oh it's not in the struct

rafiss reviewed (changes requested):
do we also need to block queries that use this data type in a cast? (that's what we had to do for `tsquery`)

we need to teach the mixed version tests to not use this type too. see 4607a7430d7 and fd09e18550a

rafiss reviewed (approved):
not gonna block this, but see my comment about cast expressions

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105391#-:-NYcvQ_rEccaxf2_0tnB:b-gwiyug)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @renatolabs, and @smg260)

___
*[`pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105391#-NYcv6R4CVE4DYyZcT2B:-NYcv6R4CVE4DYyZcT2C:b-ycx20p) ([raw file](https://github.com/cockroachdb/cockroach/blob/aad4c19b31c3a87c1f71442736e431c4dcbbe45d/pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed#L2)):*
> ```
> # LogicTest: local-mixed-22.2-23.1
> # TODO(otan): add tests for mixed 23.1-23.2.
> ```

i've got this PR brewing https://github.com/cockroachdb/cockroach/pull/104994
___
*[`pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105391#-NYcvA0pDwg9akr0mthk:-NYcvA0pDwg9akr0mthl:b-h1bipp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aad4c19b31c3a87c1f71442736e431c4dcbbe45d/pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed#L5)):*
> ```
> 
> statement error pg_lsn not supported until version 23.2
> CREATE TABLE pg_lsn_table(id pg_lsn, val pg_lsn)
> ```

i feel like we also need logic for blocking `pg_lsn` in cast expressions (and a test), otherwise won't distsql get confused in a mixed version cluster?


<!-- Sent from Reviewable.io -->

otan commented:
*[`pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105391#-NYcvA0pDwg9akr0mthk:-NYr7Yl_5wTbTd_HBHcW:b-gbfc4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/aad4c19b31c3a87c1f71442736e431c4dcbbe45d/pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed#L5)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i feel like we also need logic for blocking `pg_lsn` in cast expressions (and a test), otherwise won't distsql get confused in a mixed version cluster?
</blockquote></details>

i'll add a test; i've never had to add a mixed version test for casts before - not sure there's a great place to do that either


<!-- Sent from Reviewable.io -->

otan committed:
colinfo: add version gate for storing pg_lsn types

We don't want mixed version clusters storing pg_lsn types, in case they
need to rollback / older versions do not understand the type.

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @renatolabs, and @smg260)

___
*[`pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105391#-NYcvA0pDwg9akr0mthk:-NYrxe8j1iDma6Jim0IX:b-xwo6ll) ([raw file](https://github.com/cockroachdb/cockroach/blob/aad4c19b31c3a87c1f71442736e431c4dcbbe45d/pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed#L5)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i'll add a test; i've never had to add a mixed version test for casts before - not sure there's a great place to do that either
</blockquote></details>

you can use this same logic test file. what would be the problem with that?


<!-- Sent from Reviewable.io -->

otan commented:
*[`pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105391#-NYcvA0pDwg9akr0mthk:-NYrzul3-LaxlbSG2Dia:bcaextr) ([raw file](https://github.com/cockroachdb/cockroach/blob/aad4c19b31c3a87c1f71442736e431c4dcbbe45d/pkg/sql/logictest/testdata/logic_test/pg_lsn_mixed#L5)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

you can use this same logic test file. what would be the problem with that?
</blockquote></details>

sorry, i meant adding a version gate for "check you can't ever instantiate a pg_lsn instance anywhere distsql is used".


<!-- Sent from Reviewable.io -->

otan commented:
bors r=rafiss

mergin' for now, will update mixed-version tests later

rafiss commented:
> sorry, i meant adding a version gate for "check you can't ever instantiate a pg_lsn instance anywhere distsql is used".

you don't need to do this. all we'd need is a gate for using `pg_lsn` in cast expressions. for an example, see the change in `pkg/sql/sem/eval/cast.go` in #90842

i've added two new items to your tracking issue in #105130
>  Version gate pg_lsn type in cast expressions
> Add pgwire encoding tests for pg_lsn to pkg/cmd/generate-binary/main.go

otan commented:
> Add pgwire encoding tests for pg_lsn to pkg/cmd/generate-binary/main.go

already there!
